### ISSUE TITLE
Problem loading C# types from assemblies in debug -- cross posted from adamdriscol/poshtools

### ISSUE THREAD

**Issue by BillEidson** [microsoft/poshtools#511]
Hi, I have a ps function that loads 5 assemblies, each from specific paths. The script works in the ISE, standalone, and used to work in Visual Studio using the previous version of this tool set. The error I am trapping is as follows:

add-type : Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
At C:\Program Files\Mirth Connect\scripts\VersionIndependent\core\core.ps1:1502 char:4

add-type -Path (

```
CategoryInfo : NotSpecified: (:) [Add-Type], ReflectionTypeLoadException
FullyQualifiedErrorId : System.Reflection.ReflectionTypeLoadException,Microsoft.PowerShell.Commands.AddTypeCommand
[ERROR] Import-Module : The extension '.exe' is not a valid module extension. The supported module extensions are '.dll', '.ps1', '.psm1', '.psd1', '.cdxml' and '.xaml'. Correct the extension then try 
[ERROR] adding the file 'C:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\IDE\Extensions\mp0vfvhy.z2c\PowerShellToolsProcessHost.exe' again.
[ERROR] At line:1 char:1
[ERROR] + Import-Module 'C:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\IDE\E ...
[ERROR] + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ERROR] + CategoryInfo : PermissionDenied: (:) [Import-Module], InvalidOperationException
[ERROR] + FullyQualifiedErrorId : Modules_InvalidModuleExtension,Microsoft.PowerShell.Commands.ImportModuleCommand
[ERROR]
```

**Comment by ASayre** [comment]
Related to #438 

Hi @BillEidson ! 

Interesting script to catch this type of error; is it something you'd be willing to share to try and repro as something that would make a script break?  In any event, root cause seems related to #438.

Any chance you'd be willing to verify against the build that will come from https://ci.appveyor.com/project/PoshTools-MSFT/poshtools/build/3.0.1067 ?

**Comment by BillEidson** [comment]
He Andre,

I tried the test build you mentioned and the problem was pushed to a different place.   Now when loading assembly dll's ( add-type -Path "$global:runPath\NextGen.Core.dll" -PassThru; ) and a few more.   The assembly load fails because it cannot find dependent assemblies.   This code used to work but now that you run in a process shell, the appdomains relative search path isn't finding anything.   

Question,  how to tell it where to look?   I tried set-location and [IO.Directory]::SetCurrentDirectory to no avail.   Is it a project level option?

bille

**Comment by ASayre** [comment]
Glad the other issue is fixed!  With regards to the relative paths, '$global:runPath' is something custom :)

Can that type of behavior [if I understand what you're running into correctly] be replicated from within even standard PowerShell just by being in a different location, and then running the script by being explicit about its location?

Would using `$PSScriptRoot` in your script for relative paths fix up your issue?

**Comment by BillEidson** [comment]
I tried $PSScriptRoot.  That isn't helping the CLR load dependent assemblies.   When running a script in the poshtools debugger, the AppDomain.GetCurrent.RelativeSearchPath is pointing to the install location of the poshtools debugger.   I can't figure out how to get it to search for dependent assemblies.   This works fine running the Powershell by cmdline or the ISE.   The old poshtools also worked.

**Comment by ASayre** [comment]
That location _is_ the actual location of the process running the PowerShell context, it's not just the debugger :)  

```
Can that type of behavior [if I understand what you're running into correctly] be replicated from within even standard PowerShell just by being in a different location, and then running the script by being explicit about its location?
```

If you need to set this AppDomain value explicitly, Wouldn't https://msdn.microsoft.com/en-us/library/system.appdomainsetup.privatebinpath(v=vs.110).aspx work?

Please let us know if that fixes things up for you; I'm not understanding the bug here at this stage, so if you have more info or a repro, that would help :)

**Comment by BillEidson** [comment]
Modifying the appdomainsetup object sounds like a good way to go.  The question I have is how to get the appdomain that is running the powershell context to use it?   

I am also finding a different issue,  one of the dependent assemblies is built x86 and is failing to load because of a badimageformatexception.   This assembly is loading legacy C++ 32 bit code and is required as is.   This maybe a show stopper for me to debug this particular set of scripts in your isolated debugging enviroment for the powershell context.  Really a good thing and was hoping I wouldn't have to keep killing the IDE and restarting when it came to dot source script includes.   

any thoughts?

**Comment by ASayre** [comment]
We have an item #292 on our backlog to cover restarting the host process.  In the interim, closing out this particular issue as fixed as #438 ; With regards to your particular script, we would need more info showing an error with these tools... definitely reopen another issue if you continue to run into problems!
